lib/ask/runtime/questionnaire_action.ex

```diff
@@ -0,0 +1,245 @@
+        fn audio_ids ->
+          case audio_ids do
+            [id | tail] ->
+              audio = Repo.get_by(Audio, uuid: id)
+              {[audio], tail}
+
+            [] ->
+              {:halt, audio_ids}
+          end
+        end,
+        fn _ -> [] end
+      )
+
+    # The audio file import is based on this list
+    # If it's empty, nothing will be imported
+    audio_files =
+      Stream.map(audio_resource, fn audio ->
+        %{
+          "uuid" => audio.uuid,
+          "original_filename" => audio.filename,
+          "source" => audio.source
+        }
+      end)
+
+    audio_entries =
+      Stream.map(audio_resource, fn audio ->
+        # Zstream needs to recieve audio.data as enumerable in order to work, otherwise it throws Protocol.undefined error.
+        Zstream.entry("audios/" <> Audio.exported_audio_file_name(audio.uuid), [audio.data])
+      end)
+
+    manifest = %{
+      name: questionnaire.name,
+      modes: questionnaire.modes,
+      steps: questionnaire.steps,
+      quota_completed_steps: questionnaire.quota_completed_steps,
+      settings: questionnaire.settings,
+      partial_relevant_config: questionnaire.partial_relevant_config,
+      languages: questionnaire.languages,
+      default_language: questionnaire.default_language,
+      audio_files: audio_files
+    }
+
+    {:ok, json} = Poison.encode(manifest)
+
+    json_entry =
+      Stream.map([json], fn json ->
+        Zstream.entry("manifest.json", [json])
+      end)
+
+    zip_entries = Stream.concat(audio_entries, json_entry)
+    # since audio binary data is created as list for Zstream
+    # flatten is needed to allow the data to be sent in chunks
+    # otherwise the connection sends the whole list as a chunk
+    # and times out.
+    Zstream.zip(zip_entries)
+    |> Stream.flat_map(fn element ->
+      case is_list(element) do
+        true -> List.flatten(element)
+        false -> element
+      end
+    end)
+  end
+
+  def clean_i18n_quiz(quiz) do
+    clean_i18n_fields = Map.keys(quiz)
+
+    Enum.reduce(clean_i18n_fields, quiz, fn field, quiz_acc ->
+      clean_i18n_quiz(quiz_acc, field)
+    end)
+  end
+
+  defp clean_i18n_quiz(quiz, :settings = _field) do
+    paths = [
+      "error_message",
+      "thank_you_message"
+    ]
+
+    clean_i18n_quiz(quiz, :settings, paths)
+  end
+
+  defp clean_i18n_quiz(quiz, key) when key in [:steps, :quota_completed_steps] do
+    paths = [
+      ".[].prompt",
+      ".[].choices.[].responses.[]",
+      ".[].refusal.responses"
+    ]
+
+    clean_i18n_quiz(quiz, key, paths)
+  end
+
+  defp clean_i18n_quiz(quiz, _field) do
+    quiz
+  end
+
+  defp clean_i18n_quiz(quiz, key, paths) do
+    elem = Map.get(quiz, key)
+
+    clean_elem =
+      Enum.reduce(paths, elem, fn path, elem_acc ->
+        clean_i18n_entity(elem_acc, quiz.languages, path)
+      end)
+
+    Map.put(quiz, key, clean_elem)
+  end
+
+  def clean_i18n_entity(nil, _filter_languages, _path), do: nil
```

Comment:
all the clean_i18n_entity can be moved to its own module. They are already tested independently of the quiz, so we are half way there.

(Optional) It's worth adding a couple of comments there as an introduction in that module.